Add ESC key handling to abort WiFi connection

sideburn opened the pull request:
Checks for user ESC key input to abort the wifi connect retry loop.

sideburn committed:
Add ESC key handling to abort WiFi connection

sideburn committed:
added kbhit()

FozzTexx reviewed (changes requested):
Doesn't compile for me:

```
cl65 -t apple2 -c --create-dep obj/apple2/connect_wifi.d -Os -DBUILD_APPLE2 --include-dir fujinet-lib/4.6.2-apple2 -DUSING_FUJINET_LIB -o obj/apple2/connect_wifi.o src/connect_wifi.c
src/connect_wifi.c:86: Error: Call to undeclared function 'kbhit'
src/connect_wifi.c:88: Error: Call to undeclared function 'cgetc'
src/connect_wifi.c:89: Error: Undeclared identifier 'KEY_ESCAPE'
3 errors and 0 warnings generated.
make: *** [Makefile:322: obj/apple2/connect_wifi.o] Error 1
```

cc65 --version
cc65 V2.19 - Git d7d1d8969

sideburn committed:
Added required includes, changed ESC key check to 'x' to bypass need for key_codes

FozzTexx commented on src/connect_wifi.c:
```diff
@@ -82,6 +82,19 @@ void connect_wifi(void)
+		if (kbhit()) 
+		{
+			key = cgetc();
+			if (key == 'x')
```
This is looking for the lowercase `x` key, not ESC.

sideburn commented:
I changed it to use the X key instead son it’s universal across all platforms and doesn’t need to load the key-mapping filesOn Apr 4, 2025, at 7:15 AM, FozzTexx ***@***.***> wrote:﻿
@FozzTexx requested changes on this pull request.



In src/connect_wifi.c:
> @@ -82,6 +82,19 @@ void connect_wifi(void)
 
 	while (retries > 0)
 	{
+		// check for x key and abort
+		if (kbhit()) 
+		{
+			key = cgetc();
+			if (key == 'x')

This is looking for the lowercase x key, not ESC.

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you authored the thread.Message ID: ***@***.***>

tschak909 commented:
> I changed it to use the X key instead son it’s universal across all platforms and doesn’t need to load the key-mapping filesOn Apr 4, 2025, at 7:15 AM, FozzTexx ***@***.***> wrote:﻿ @FozzTexx requested changes on this pull request. In src/connect_wifi.c:
> @@ -82,6 +82,19 @@ void connect_wifi(void)
> while (retries > 0) { + // check for x key and abort + if (kbhit()) + { + key = cgetc(); + if (key == 'x') This is looking for the lowercase x key, not ESC. —Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you authored the thread.Message ID: ***@***.***>

When checking for a key that can do both upper and lower case, please check for both upper and lowercase.

sideburn commented:
It can just be lowercase.. Any key to do an abort is fine.. I changed it to an ‘x’ so it would be cross platform without needing to include they key_codes.h file which becomes problematic with macro redefinition errors and also looks like the esc key mapping is only being used for c64 and the key_codes is only being included in edit_strinig.c so I didn't want to risk having issues.

> On Apr 4, 2025, at 9:26 AM, Thomas Cherryhomes ***@***.***> wrote:
> 
> 
> tschak909
>  left a comment 
> (FujiNetWIFI/fujinet-config#113)
> I changed it to use the X key instead son it’s universal across all platforms and doesn’t need to load the key-mapping filesOn Apr 4, 2025, at 7:15 AM, FozzTexx @.> wrote:﻿ @FozzTexx requested changes on this pull request. In src/connect_wifi.c:
> @@ -82,6 +82,19 @@ void connect_wifi(void)
> while (retries > 0) { + // check for x key and abort + if (kbhit()) + { + key = cgetc(); + if (key == 'x') This is looking for the lowercase x key, not ESC. —Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you authored the thread.Message ID: @.>
> 
> When checking for a key that can do both upper and lower case, please check for both upper and lowercase.
> 
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you authored the thread.
> 
>  <https://github.com/FozzTexx> <https://github.com/FujiNetWIFI/fujinet-config/pull/113#issuecomment-2779232466> <https://github.com/notifications/unsubscribe-auth/ABXCICUUFJ4N6QBO45UBTTD2X2XEVAVCNFSM6AAAAABZ4DCAASVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDONZZGIZTENBWGY>
> 
> tschak909
>  left a comment 
> (FujiNetWIFI/fujinet-config#113)
>  <https://github.com/FujiNetWIFI/fujinet-config/pull/113#issuecomment-2779232466>
> I changed it to use the X key instead son it’s universal across all platforms and doesn’t need to load the key-mapping filesOn Apr 4, 2025, at 7:15 AM, FozzTexx @.> wrote:﻿ @FozzTexx <https://github.com/FozzTexx> requested changes on this pull request. In src/connect_wifi.c:
> @@ -82,6 +82,19 @@ void connect_wifi(void)
> while (retries > 0) { + // check for x key and abort + if (kbhit()) + { + key = cgetc(); + if (key == 'x') This is looking for the lowercase x key, not ESC. —Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you authored the thread.Message ID: @.>
> 
> When checking for a key that can do both upper and lower case, please check for both upper and lowercase.
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/FujiNetWIFI/fujinet-config/pull/113#issuecomment-2779232466>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ABXCICUUFJ4N6QBO45UBTTD2X2XEVAVCNFSM6AAAAABZ4DCAASVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDONZZGIZTENBWGY>.
> You are receiving this because you authored the thread.
>

sideburn committed:
Changed abort key code from 'x' to ESC

sideburn committed:
Use KCODE_ESCAPE for abort. Updated c64 header.

sideburn committed:
Added KCODEs to Apple and Adam

FozzTexx commented on src/apple2/screen.h:
```diff
@@ -71,5 +71,10 @@ void screen_select_slot_mode(void);
+
+#define KCODE_RETURN 0x9B // is the ATASCI equivlant of 155 End Of Line (return)
+#define KCODE_ESCAPE 0x1B
+#define KCODE_BACKSP 0x7E
```
0x7E is the code for `~` on Apple IIe (and later models), not backspace. Backspace (left arrow) is 0x08 and DELETE is 0x7F. RETURN is 0x0D, not 0x9B.

FozzTexx commented on src/c64/screen.h:
```diff
@@ -69,5 +69,13 @@ void screen_select_slot(char *e);
+ */
+#define KCODE_RETURN 0x9B 
+#define KCODE_ESCAPE 0x1B
+#define KCODE_BACKSP 0x7E
```
C64 doesn't have an ESC or backspace key. It does have a key labeled ← where the ESC is on other computers. The ← key is also different than the cursor keys ⇐ ⇑ ⇒ ⇓

sideburn commented:
The left arrow key, located in the upper left corner of the c64 keyboard is interpreted as the ESC key. 

> On Apr 11, 2025, at 9:38 AM, FozzTexx ***@***.***> wrote:
> 
> 
> @FozzTexx commented on this pull request.
> 
> In src/apple2/screen.h <https://github.com/FujiNetWIFI/fujinet-config/pull/113#discussion_r2039895989>:
> 
> > @@ -71,5 +71,10 @@ void screen_select_slot_mode(void);
>  void screen_select_slot_eject(unsigned char ds);
>  
>  bool screen_mount_and_boot_lobby(void);
> +
> +#define KCODE_RETURN 0x9B // is the ATASCI equivlant of 155 End Of Line (return)
> +#define KCODE_ESCAPE 0x1B
> +#define KCODE_BACKSP 0x7E
> 0x7E is the code for ~ on Apple IIe (and later models), not backspace. Backspace (left arrow) is 0x08 and DELETE is 0x7F.
> 
> In src/c64/screen.h <https://github.com/FujiNetWIFI/fujinet-config/pull/113#discussion_r2039902474>:
> 
> > +#define KCODE_ESCAPE 0x1B
> +#define KCODE_BACKSP 0x7E
> C64 doesn't have an ESC or backspace key. It does have a key labeled ← where the ESC is on other computers. The ← key is also different than the cursor keys ⇐ ⇑ ⇒ ⇓
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/FujiNetWIFI/fujinet-config/pull/113#pullrequestreview-2760911371>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ABXCICXXFWLFGMZGLYGRAPL2Y7VZHAVCNFSM6AAAAABZ4DCAASVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDONRQHEYTCMZXGE>.
> You are receiving this because you authored the thread.
>

sideburn committed:
Reverted additional unused KCODES from Adam,Apple,c64, only added ESC Keycodes. C64's top left key label is a left arrow but keycode is still ESC code.

FozzTexx commented:
> The left arrow key, located in the upper left corner of the c64 keyboard is interpreted as the ESC key.

It might be in the location where ESC is on other keyboards, but it isn't labeled as ESC and it doesn't send 0x1B.

sideburn commented:
So what do you propose as a solution? Remove the ability to abort when on a C64? Provide ESC labels to stick over the key caps? if we don't want to make that the abort key (change to 0Xf5 or whatever) then I am not sure how to move forward. It's taking two weeks to get this tiny itty bitty feature in lol


> On Apr 11, 2025, at 2:13 PM, FozzTexx ***@***.***> wrote:
> 
> 
> FozzTexx
>  left a comment 
> (FujiNetWIFI/fujinet-config#113)
> The left arrow key, located in the upper left corner of the c64 keyboard is interpreted as the ESC key.
> 
> It might be in the location where ESC is on other keyboards, but it isn't labeled as ESC and it doesn't send 0x1B.
> 
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you authored the thread.
> 
>  <https://github.com/FujiNetWIFI/fujinet-config/pull/113#issuecomment-2798020763> <https://github.com/notifications/unsubscribe-auth/ABXCICW6R7OAYLSGRK7FRLL2ZAV6HAVCNFSM6AAAAABZ4DCAASVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDOOJYGAZDANZWGM>
> 
> FozzTexx
>  left a comment 
> (FujiNetWIFI/fujinet-config#113)
>  <https://github.com/FujiNetWIFI/fujinet-config/pull/113#issuecomment-2798020763>
> The left arrow key, located in the upper left corner of the c64 keyboard is interpreted as the ESC key.
> 
> It might be in the location where ESC is on other keyboards, but it isn't labeled as ESC and it doesn't send 0x1B.
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/FujiNetWIFI/fujinet-config/pull/113#issuecomment-2798020763>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ABXCICW6R7OAYLSGRK7FRLL2ZAV6HAVCNFSM6AAAAABZ4DCAASVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDOOJYGAZDANZWGM>.
> You are receiving this because you authored the thread.
>

FozzTexx commented:
> So what do you prose as a solution? Remove the ability to abort when on a C64? Provide ESC labels to stick over the key caps?

Using a sharpie to write ESC on the left arrow key won't make it start sending 0x1B. The `#define` needs to use the correct key code and the correct key name so as to prevent confusion about what key the `#define` represents.

Could `#define` a KEY_ABORT at the top of `connect_wifi.c` and put the necessary #ifdef stuff there. Probably a matching KEY_ABORT_LABEL or something for the message text.

sideburn commented:
> > So what do you prose as a solution? Remove the ability to abort when on a C64? Provide ESC labels to stick over the key caps?
> 
> Using a sharpie to write ESC on the left arrow key won't make it start sending 0x1B. The `#define` needs to use the correct key code and the correct key name so as to prevent confusion about what key the `#define` represents.
> 
> Could `#define` a KEY_ABORT at the top of `connect_wifi.c` and put the necessary #ifdef stuff there. Probably a matching KEY_ABORT_LABEL or something for the message text.

I had #defines and took them out because theres already screen.h and is already ifdefed for each platform. the c64 code for "ESC" is 0Xf5.. I will rework the defines and fix the code if you are ok with the top left <-- key being the "abort" key for c64

sideburn commented:
> > So what do you prose as a solution? Remove the ability to abort when on a C64? Provide ESC labels to stick over the key caps?
> 
> Using a sharpie to write ESC on the left arrow key won't make it start sending 0x1B. The `#define` needs to use the correct key code and the correct key name so as to prevent confusion about what key the `#define` represents.
> 
> Could `#define` a KEY_ABORT at the top of `connect_wifi.c` and put the necessary #ifdef stuff there. Probably a matching KEY_ABORT_LABEL or something for the message text.

is this what you prefer:

#ifndef BUILD_C64
      if (key == KCODE_ESCAPE)
#else
     if (key == KCODE_ABORT)
#endif

c64/screen.h:

#define KCODE_ABORT 0x5F

#endif /* SCREEN_H */
#endif /* BUILD_C64 */

sideburn commented:
> > So what do you prose as a solution? Remove the ability to abort when on a C64? Provide ESC labels to stick over the key caps?
> 
> Using a sharpie to write ESC on the left arrow key won't make it start sending 0x1B. The `#define` needs to use the correct key code and the correct key name so as to prevent confusion about what key the `#define` represents.
> 
> Could `#define` a KEY_ABORT at the top of `connect_wifi.c` and put the necessary #ifdef stuff there. Probably a matching KEY_ABORT_LABEL or something for the message text.

** if they all used the same KCODE_ESCAPE define then there be less code and don't need the #ifndef...  Would just mean ESC means ABORT on c64.  its even called escape already in key_codes.h

// back arrow <-
#define KEY_ESCAPE         0x5F

Personally I prefer to not add code if its not needed.

sideburn committed:
FIxed incorrect ESC code for c64 same as currently defined KCODE_ESCAPE 0x5F in key_codes.h for c64

FozzTexx commented:
> ** if they all used the same KCODE_ESCAPE define then there be less code 

It may be "less code" but it's important that the code also be readable and maintainable by the next person that works on it. When someone working on the C64 side of it sees `KEY_ESCAPE` they are going to wonder what key that is since there isn't one of those on the C64.

A more maintainable approach is to define the constant with the actual key name (e.g., `KEY_LEFTARROW`) and then create a separate #define like `ABORT_KEY` which uses the previously defined constants. That way, `ABORT_KEY` expresses the intention, and the mapping to an available key stays explicit and easy to follow. The `ABORT_KEY` `#define` would be wrapped with #ifdefs for cross-platform compatibility without creating confusion.

I know that the left arrow key has already been used as an ESC substitute on the C64, so it's fine to use that key for now. But I know from experience that when something says to push ← that it can be confusing to find that key because it's in an odd location and because there are cursor keys that also have arrows on them. I think down the road we are going to want to change the keys on the C64 to use things that are easier for users to identify, such as F1, F3, F5, F7. Making an `ABORT_KEY` `#define` instead of mislabeling the `#define` to match the same name as other platforms will make it easier to swap it from `KEY_LEFTARROW` to `KEY_F7`.

sideburn committed:
Enabled for A8 build only.

tschak909 committed:
properly refactor input handling for abort.

tschak909 committed:
Merge branch 'main' into add-escape-key-to-wifi-connect

tschak909 commented:
OK, I've gone through and refactored things a bit:

@sideburn Because we're targeting a whole bunch of compilers, each with their own libraries, some things have conio, some don't, but we still need to abstract things, so e.g. we have an input() method that's in src/[xxx]/input that does all the heavy lifting. it returns a key if one is present, otherwise it doesn't block.

So I've changed the code to use that. and have added a KEY_ABORT to the input.h of all of the platforms to match.

sideburn commented:
I like it. On same topic, I happened to be reading the Sublogic Flight Simulator manual yesterday and where they mention pressing ESC key it says (E key on c64)… found that interesting…On May 21, 2025, at 5:51 AM, Thomas Cherryhomes ***@***.***> wrote:﻿
tschak909 left a comment (FujiNetWIFI/fujinet-config#113)
OK, I've gone through and refactored things a bit:
@sideburn Because we're targeting a whole bunch of compilers, each with their own libraries, some things have conio, some don't, but we still need to abstract things, so e.g. we have an input() method that's in src/[xxx]/input that does all the heavy lifting. it returns a key if one is present, otherwise it doesn't block.
So I've changed the code to use that. and have added a KEY_ABORT to the input.h of all of the platforms to match.

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you were mentioned.Message ID: ***@***.***>
